BAH 1556 | Add crater-atomfeed initital version

TuhinBeehyv opened the pull request:
Initial version of crater-atomfeed

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Bahmni/crater-atomfeed?pullRequest=1) <br/>All committers have signed the CLA.

N0-man commented on feed-integration-webapp/pom.xml:
```diff
@@ -0,0 +1,357 @@
+                    <target>11</target>
+                </configuration>
+            </plugin>
+            <!--            <plugin>-->
```
remove

N0-man commented on feed-integration-webapp/pom.xml:
```diff
@@ -0,0 +1,357 @@
+                    <groupId>ch.qos.logback</groupId>
+                    <artifactId>logback-classic</artifactId>
+                </exclusion>
+                <!--                <exclusion>-->
```
remove all commented code

N0-man commented on feed-integration-webapp/pom.xml:
```diff
@@ -0,0 +1,357 @@
+            <scope>runtime</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.h2database</groupId>
```
are we using h2?

N0-man commented on feed-integration-webapp/pom.xml:
```diff
@@ -0,0 +1,357 @@
+        <dependency>
+            <groupId>org.json</groupId>
+            <artifactId>json</artifactId>
+            <version>20090211</version>
```
why use such an old version?
Can you check all the dependency versions and take it to latest?

N0-man commented on feed-integration-webapp/pom.xml:
```diff
@@ -0,0 +1,357 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Why use maven? We could very well use gradle which is way more performant and overall much better

N0-man commented on feed-integration-webapp/src/main/java/org/bahmni/module/CraterAtomfeed.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+
+//    @Override
```
remove

N0-man commented on feed-integration-webapp/src/main/java/org/bahmni/module/CraterAtomfeed.java:
```diff
@@ -0,0 +1,59 @@
+        auth = crater.login();
+
+        SpringApplication.run(CraterAtomfeed.class, args);
+        System.out.println("************ Started Crater Atomfeed app **********");
```
remove the sysout

N0-man commented on feed-integration-webapp/src/main/java/org/bahmni/module/CraterAtomfeed.java:
```diff
@@ -0,0 +1,59 @@
+
+    @RequestMapping("/")
+    String home() {
+        return "PACS Integration module is up and running.";
```
PACS integration? I believe this home endpoint is for health check

N0-man commented on feed-integration-webapp/src/main/java/org/bahmni/module/feedintegration/apicalls/CraterAPIcalls.java:
```diff
@@ -0,0 +1,238 @@
+            con.setRequestProperty("Accept", properties.getProperty("con.accept"));
+            con.setRequestProperty("company", properties.getProperty("crater.company"));
+            con.setDoOutput(Boolean.parseBoolean(properties.getProperty("con.setDoOutput")));
+            con.setDoInput(Boolean.parseBoolean(properties.getProperty("con.setDoInput")));
```
the code is DRY for create and update. You can refactor with method extraction

```
private HttpURLConnection getConnection(String api, String httpMethod, Properties properties) throws IOException {
        URL url = new URL(properties.getProperty("crater.url") + api);
        HttpURLConnection con = (HttpURLConnection) url.openConnection();
        con.setRequestMethod(httpMethod);
        con.setRequestProperty("Authorization", "Bearer " + auth);
        con.setRequestProperty("Content-Type", properties.getProperty("con.content_type"));
        con.setRequestProperty("Accept", properties.getProperty("con.accept"));
        con.setRequestProperty("company", properties.getProperty("crater.company"));
        con.setDoOutput(Boolean.parseBoolean(properties.getProperty("con.setDoOutput")));
        con.setDoInput(Boolean.parseBoolean(properties.getProperty("con.setDoInput")));
        return con;
    }
```

and then 
```
HttpURLConnection con = getConnection("customers", "POST", properties);
```

N0-man commented on feed-integration-webapp/src/main/java/org/bahmni/module/feedintegration/apicalls/CraterAPIcalls.java:
```diff
@@ -0,0 +1,238 @@
+
+    String auth = CraterAtomfeed.auth;
+
+    public Properties getprop() throws IOException{
```
properties could be a singleton dependency or a private construct 

```
public CraterAPIcalls() throws IOException {
        auth = CraterAtomfeed.auth;
        properties = getProps();
    }
```
and then the `CraterAPIcalls` class would have `Properties properties;`

N0-man commented on feed-integration-webapp/src/main/java/org/bahmni/module/feedintegration/apicalls/CraterAPIcalls.java:
```diff
@@ -0,0 +1,238 @@
+
+    }
+
+    public String del_customer(String auth, OpenMRSPatientFullRepresentation patient) throws JSONException, IOException, URISyntaxException {
```
unused code... please delete

N0-man commented on feed-integration-webapp/src/main/java/org/bahmni/module/feedintegration/apicalls/CraterAPIcalls.java:
```diff
@@ -0,0 +1,238 @@
+    }
+
+    public void create_update(OpenMRSPatientFullRepresentation patientFR) throws Exception {
+        if (login_verification(auth) == "Authenticated") {
```
`login_verification(auth).equals("Authenticated")`

N0-man commented on feed-integration-webapp/src/main/java/org/bahmni/module/feedintegration/apicalls/CraterAPIcalls.java:
```diff
@@ -0,0 +1,238 @@
+                break;
+            }
+        }
+        return uuid;
```
imperative code with mutations... better leverage language features and increase the readability
```
      Optional<OpenMRSPatientIdentifier> uuid1 = patientFR.getIdentifiers().stream()
                .filter(OpenMRSPatientIdentifier::isPreferred)
                .findFirst();

        return uuid1.isPresent() ? uuid1.get().getIdentifier() : EMPTY_STRING;
```

N0-man commented on feed-integration-webapp/src/main/java/org/bahmni/module/feedintegration/apicalls/CraterAPIcalls.java:
```diff
@@ -0,0 +1,238 @@
+                    + " " + patientFR.getPerson().getPreferredName().getMiddleName()
+                    + " " + patientFR.getPerson().getPreferredName().getFamilyName();
+        }
+        return name;
```
imperative and DRY...

```
OpenMRSPersonName preferredName = patientFR.getPerson().getPreferredName();
return preferredName.getGivenName()
        + (preferredName.getMiddleName() != null ? " " + preferredName.getMiddleName() : EMPTY_STRING)
        + " " + preferredName.getFamilyName();
```

N0-man reviewed (commented):
Its a huge PR - need more time to review. Sharing some initial comments to look at

N0-man commented:
@TuhinBeehyv and I had a round of discussion - In addition to the comments there is a lot of unnecessary and unused code that sneaked from the  [bahmni-feed-integration-example](https://github.com/Bahmni/bahmni-feed-integration-example). @TuhinBeehyv will clean it up

atish160384 commented on feed-integration-webapp/src/main/java/org/bahmni/module/feedintegration/crater/CraterAPIcalls.java:
```diff
@@ -0,0 +1,158 @@
+public class CraterAPIcalls {
+
+    private final CraterProperties properties = CraterProperties.getInstance();
+    CraterLogin craterLogin = new CraterLogin();
```
@Autowire

N0-man reviewed (commented):
Few things to consider
- Move from Maven to Gradle
- `PACS_CONCEPT_SOURCE_NAME` constant is bogus
- In liquibase, the auther name could be `Bahmni` (having individual developers name is irrelevant)
- log4J root level should not be DEBUG
- remove OpenMRS copyright notice from HL7Utils
- `scripts/datasetup/pg-initdb.sh` is bogus
- refactor `setupDB.sql` to remove any organization specifics (beehyv).. responsibility of creating of user and database would anyways be outside of application scope

TuhinBeehyv commented:
1. Maven to Gradle would take a lot of time so its better to do it on a separate ticket later as its not high priority
2. File has been removed as its not relevant
3. Changed to `bahmni`
4. Changed to `INFO`
5. Removed
6. File has been removed as its not relevant
7. File has been removed as its not relevant

gsluthra commented on feed-integration-webapp/src/main/java/org/bahmni/module/feedintegration/atomfeed/client/AtomFeedProperties.java:
```diff
@@ -0,0 +1,72 @@
+
+        } catch (Exception e) {
+        } finally {
+            if (null != propertyStream) {
```
This line is not needed. Even if a null pointer exception is thrown in the next line, it will get caught and ignored.

TuhinBeehyv replied:
Line has been removed.

gsluthra commented on feed-integration-webapp/src/main/java/org/bahmni/module/feedintegration/atomfeed/contract/patient/OpenMRSPerson.java:
```diff
@@ -0,0 +1,167 @@
+    private String gender;
+    private int age;
+    private Date birthdate;
+    private String birthtime;
```
Why do we need all this data about a person in crater integration code? For creating a Customer, my guess is we only need minor details:  Name, Age, Gender, Patient ID, Address.

Other information here like birthtime, death details, preferredNames, etc... should not be needed. Can we not ignore those while constructing the OpenMRSPerson object? Else, more personal data is travelling across services.

TuhinBeehyv replied:
Irrelevant info has been removed.

gsluthra commented on feed-integration-webapp/src/main/java/org/bahmni/module/feedintegration/atomfeed/mappers/OpenMRSPatientMapper.java:
```diff
@@ -0,0 +1,38 @@
+        OpenMRSPatient patient = new OpenMRSPatient();
+        JsonNode jsonNode = objectMapper.readTree(patientJSON);
+
+        patient.setPatientId(jsonNode.path("identifiers").get(0).path("identifier").asText());
```
The mapper here seems to be correct. Uses minimal information. Maybe its just the OpenMRS Person/Patient object then that needs to drop the extra getter/setters from its definition.

gsluthra commented on feed-integration-webapp/src/main/java/org/bahmni/module/feedintegration/crater/CraterAPIcalls.java:
```diff
@@ -0,0 +1,158 @@
+        return con;
+    }
+
+    public String login_verification(String auth) throws Exception {
```
Why does this method return Strings, instead of a type or boolean? 
I would recommend to choose boolean to indicate login successful/failed, or use an enum.

TuhinBeehyv replied:
Changed it to boolean.

gsluthra commented on feed-integration-webapp/src/main/java/org/bahmni/module/feedintegration/crater/CraterAPIcalls.java:
```diff
@@ -0,0 +1,158 @@
+
+    public String getname(OpenMRSPatientFullRepresentation patientFR) {
+        OpenMRSPersonName preferredName = patientFR.getPerson().getPreferredName();
+        return preferredName.getGivenName()
```
This logic should be in OpenMRSPersonName class with a method called: getPreferredFullName( ).

gsluthra commented on feed-integration-webapp/src/main/resources/atomfeed.properties:
```diff
@@ -0,0 +1,15 @@
+#OpenMRS data specific properties
+openmrs.auth.uri=http://localhost:8080/openmrs/ws/rest/v1/session
+openmrs.patient.feed.uri=http://localhost:8080/openmrs/ws/atomfeed/patient/recent
+openmrs.user=superman
```
Crater bridge should not use `superman` to talk to openmrs. Best to use a crater specific credentials string. We don't want crater calls to look like these were "user" made calls to OpenMRS.

TuhinBeehyv replied:
Waiting on response from Mohan about how a new user can be created automatically when running the crater-atomfeed docker image with bahmni docker image.

gsluthra commented on feed-integration-webapp/src/main/resources/crater.properties:
```diff
@@ -0,0 +1,11 @@
+crater.url = https://payments-dev.lite.mybahmni.in
+crater.company = 1
+crater.device_name = mobile_app
+crater.username = superman@bahmni.org
```
Use a bridge specific credential, rather than using a generic superadmin credential.

TuhinBeehyv replied:
Similar to above issue, I have asked Venu for the credentials.

gsluthra commented on feed-integration-webapp/src/main/resources/db/changelog/liquibase.xml:
```diff
@@ -0,0 +1,50 @@
+    <changeSet id="7" author="Bahmni">
+        <comment>Adding schedulers for openmrs patient</comment>
+        <sql>
+            insert into quartz_cron_scheduler (name, cron_statement, start_delay, enabled) values ('openMRSPatientFeedJob', '0/15 * * * * ?', 0, true);
```
Suggest to rename this cron job to: `openmrsPatientFeedForCraterJob` to ensure its clear why this cron job is registered in the system.

gsluthra replied:
It should ideally run every 2 mins. 15 mins seems to long a time. For clinics we don't expect too much foot fall, so the scheduler should finish quite quickly. We increase this time only if this impacts system performance.

TuhinBeehyv replied:
The cron job is currently running every 15secs, not 15 mins. Let me know if it should run every 2 mins instead. Name of cron job has been changed.

gsluthra commented on feed-integration-webapp/src/test/resources/samplePatient.json:
```diff
@@ -0,0 +1,192 @@
+  ],
+  "person": {
+    "uuid": "31e164fa-dd5b-4a10-a487-33e3a97e6198",
+    "display": "Test NoOpenVisit Scenario",
```
Is there a reason why this JSON doesn't look like an actual Patient json? It would be better to put real sounding names to ensure the test acts as actual documentation on what is expected. Having a given name with semi-colons is invalid.

TuhinBeehyv replied:
A real patient name has been added for test files.

gsluthra commented on feed-integration-webapp/src/test/java/org/bahmni/module/feedintegration/atomfeed/mappers/OpenMRSPatientMapperTest.java:
```diff
@@ -0,0 +1,21 @@
+        OpenMRSPatient patient = new OpenMRSPatientMapper().map(json);
+
+        Assert.assertEquals("GAN200053", patient.getPatientId());
+        Assert.assertEquals(" T e s t N o O p e_n-V i s i t ", patient.getGivenName());
```
Please make the assertions realistic. TestOpenVisit is a confusing Patient name.

gsluthra commented on feed-integration-webapp/src/test/java/org/bahmni/module/feedintegration/integrationtest/BaseIntegrationTest.java:
```diff
@@ -0,0 +1,60 @@
+    private static final Logger log = LoggerFactory.getLogger(BaseIntegrationTest.class);
+
+    @Before
+    public void startModalityStubServer() throws InterruptedException, UnknownHostException {
```
What is this method for? Modality is PACS terminology. Why is this here?

TuhinBeehyv replied:
Test was not being used so has been removed.

gsluthra commented on feed-integration-webapp/src/test/java/org/bahmni/module/feedintegration/integrationtest/HL7Utils.java:
```diff
@@ -0,0 +1,54 @@
+import java.text.*;
+import java.util.*;
+
+public class HL7Utils {
```
Is this class being used for CraterAtom feed?

TuhinBeehyv replied:
Class/Test was not being used so has been removed.

gsluthra commented on feed-integration-webapp/src/main/java/org/bahmni/module/feedintegration/crater/CraterAPIcalls.java:
```diff
@@ -0,0 +1,158 @@
+                byte[] inputString = jsonInputString.getBytes(UTF_8);
+                os.write(inputString, 0, inputString.length);
+            }
+            con.getInputStream();
```
Don't these streams need to be closed after the create_customer( ) operation is done? Or do these remain open throughout the lifecycle of this class?

TuhinBeehyv replied:
Streams have been closed. They don't need to remain open.

gsluthra commented on feed-integration-webapp/src/main/java/org/bahmni/module/feedintegration/crater/CraterAPIcalls.java:
```diff
@@ -0,0 +1,158 @@
+import static com.sun.org.apache.xml.internal.utils.LocaleUtility.EMPTY_STRING;
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+public class CraterAPIcalls {
```
Instead of calling this class `CraterAPIcalls` can we rename it to something like: `CraterAPIClient` or `CraterWrapper` or just `Crater`?

TuhinBeehyv replied:
Name changed to CraterAPIClient.

gsluthra commented on feed-integration-webapp/src/main/java/org/bahmni/module/feedintegration/crater/CraterAPIcalls.java:
```diff
@@ -0,0 +1,158 @@
+public class CraterAPIcalls {
+
+    private final CraterProperties properties = CraterProperties.getInstance();
+    CraterLogin craterLogin = new CraterLogin();
```
I would suggest to move this into the constructor. Can't this be autowired if it is Spring Managed singleton?

TuhinBeehyv replied:
craterLogin has been autowired.

gsluthra commented on README.md:
```diff
@@ -0,0 +1,16 @@
+# Crater Atomfeed
+
```
Please provide a proper description of this project for newbies, and mention the entry point of the code, for them to understand how to read this project.

gsluthra commented on README.md:
```diff
@@ -0,0 +1,16 @@
+* mvn 3.6.2. 
+* Java 1.8, works on 11 as well (with some warnings) 
+
+####  running the app 
```
Please mention how to run tests.

gsluthra commented on feed-integration-webapp/src/main/resources/application.properties:
```diff
@@ -0,0 +1,13 @@
+spring.datasource.url=jdbc:mysql://localhost:3306/crater_atomfeed
+spring.datasource.username=beehyv
```
Why does this say beehvy?

TuhinBeehyv replied:
Changed to user/{no password}

angshu commented on .classpath:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Guess this is a eclipse IDE file, and I would prefer not to add this to repo. Can you please also add to gitignore

TuhinBeehyv replied:
Folder has been removed

angshu commented on .project:
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Guess this is a eclipse IDE file, and I would prefer not to add this to repo. Can you please also add to gitignore

TuhinBeehyv replied:
Folder has been removed

angshu commented on feed-integration-webapp/src/main/java/org/bahmni/module/feedintegration/CraterAtomfeed.java:
```diff
@@ -0,0 +1,33 @@
+import org.springframework.transaction.annotation.EnableTransactionManagement;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
```
Do we need to annotate with rest controller. this is just a app initializer. so just springBootApplication annotation should be enough?

angshu replied:
also, we dont need a controller (either using webmvc or rest controller unless we are exposing any api.

TuhinBeehyv replied:
It's no longer a restcontroller

angshu commented on feed-integration-webapp/src/main/java/org/bahmni/module/feedintegration/CraterAtomfeed.java:
```diff
@@ -0,0 +1,33 @@
+@SpringBootApplication
+@ComponentScan(basePackages = "org.bahmni.module.*")
+@EnableTransactionManagement
+public class CraterAtomfeed extends SpringBootServletInitializer {
```
also by convention the main application class is probably better named as simple "Application" (instead of CraterAtomfeed)

TuhinBeehyv replied:
Name has been changed to Application

angshu commented on feed-integration-webapp/src/main/java/org/bahmni/module/feedintegration/CraterAtomfeed.java:
```diff
@@ -0,0 +1,33 @@
+
+@RestController
+@SpringBootApplication
+@ComponentScan(basePackages = "org.bahmni.module.*")
```
Are you scanning only this project's directories? this will make spring scan even compiled and bundled jars. Better to be more specific "org.bahmni.module.feedintegration.*"

TuhinBeehyv replied:
Changed to org.bahmni.module.feedintegration.

angshu commented on feed-integration-webapp/src/main/java/org/bahmni/module/feedintegration/CraterAtomfeed.java:
```diff
@@ -0,0 +1,33 @@
+        logger.info("************ Started Crater Atomfeed app **********");
+    }
+
+    @Bean
```
these are better configured in a separate java-config class (e.g. Configuration.java).. annotated with @configuration

angshu commented on feed-integration-webapp/src/main/java/org/bahmni/module/feedintegration/atomfeed/ScheduledTasks.java:
```diff
@@ -0,0 +1,86 @@
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+@Configuration
```
prefer moving them to a single configuration class

angshu commented on feed-integration-webapp/src/main/java/org/bahmni/module/feedintegration/crater/CraterAPIClient.java:
```diff
@@ -0,0 +1,159 @@
+
+    private static final Logger logger = LoggerFactory.getLogger(CraterAPIClient.class);
+
+    public HttpURLConnection getConnection(String api, String httpMethod, String auth) throws IOException {
```
Does this method need to be public? this is used internally .. so can be private.

TuhinBeehyv replied:
Has been made private

angshu commented on feed-integration-webapp/src/main/java/org/bahmni/module/feedintegration/atomfeed/client/AtomFeedProperties.java:
```diff
@@ -0,0 +1,70 @@
+
+    private static AtomFeedProperties atomFeedProperties;
+
+    private AtomFeedProperties() {
```
We are loading from classpath application.properties .. is this a good idea? why not use a PropertyPlaceholderConfigurer or leverage spring externalized configuration support, so that anyone can override using either -D or using system variables or even using "--spring.config.location" at startup.

angshu replied:
I think you can just use @Value("${property-name}") and when a property source is configured .. it will automatically do the needful

TuhinBeehyv replied:
There were 2 such properties files, CraterProperties and AtomfeedProperties. CraterProperties was made to keep in line with how property files were being imported. I have changed it back to using @Value annotation. But unfortunately that can't be done to AtomfeedProperties as the whole class is being used in methods inside atomfeed dependency, which I can't edit. I also tried to use @Value to take a variable property name (just like how getProperty does) but @Value can't take variables thus doesn't work. So Properties related to Crater have been imported using @Value annotation in spring but AtomfeedProperties will have to stay the same unless atomfeed dependency is changed to not need that particular class.

angshu commented on feed-integration-webapp/src/main/resources/application.properties:
```diff
@@ -0,0 +1,13 @@
+spring.datasource.url=jdbc:mysql://localhost:3306/crater_atomfeed
```
default application properties are good idea, however without a way to override them it will make things difficult for someone to change - even for other properties file that may require dynamic property. 

You should consider externalizing configuration .. use a propertysourceconfigurer, use @value('property-name') and leverage using -D parameter, or system variable or even  either pass at startup "--spring.config.location=classpath:/another-location.properties"

The above leaves options open and does not require code change and allows docker or other environments to manage them externally and pass them to the app using whatever means suits them

angshu commented on feed-integration-webapp/src/main/resources/atomfeed.properties:
```diff
@@ -0,0 +1,16 @@
+#Server properties
```
Same comment as above

angshu commented on feed-integration-webapp/src/main/resources/crater.properties:
```diff
@@ -0,0 +1,11 @@
+con.setDoInput = true
+
+crater.currencyid = 13 #for INR
+crater.url = https://payments-dev.lite.mybahmni.in
```
same comment about externalizing configuration .. these should be configurable externally ..

angshu commented on feed-integration-webapp/src/main/resources/crater.properties:
```diff
@@ -0,0 +1,11 @@
+crater.company = 1
+crater.device_name = mobile_app
+crater.username = superman@bahmni.org
+crater.password = Crater123
```
no password in properties file (unless for test). such property (username/passwords/urls) should be externalized.

angshu commented on feed-integration-webapp/src/main/java/org/bahmni/module/feedintegration/Application.java:
```diff
@@ -25,9 +26,19 @@
+import org.springframework.transaction.annotation.EnableTransactionManagement;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
```
Why restcontroller?

angshu commented on feed-integration-webapp/src/main/java/org/bahmni/module/feedintegration/Application.java:
```diff
@@ -25,9 +26,19 @@
+
+@RestController
+@SpringBootApplication
+@ComponentScan(basePackages = "org.bahmni.module.*")
```
limit scan to "org.bahmni.module.feedintegration.*" at the least

angshu commented on feed-integration-webapp/src/main/java/org/bahmni/module/feedintegration/crater/CraterAPIClient.java:
```diff
@@ -16,6 +16,8 @@
+import org.springframework.context.annotation.PropertySource;
 import org.springframework.stereotype.Component;
 
 import java.io.*;
```
no package.* import please

angshu commented on feed-integration-webapp/src/main/java/org/bahmni/module/feedintegration/crater/CraterLogin.java:
```diff
@@ -17,19 +19,30 @@
     public CraterLogin() {
         try {
-            CraterProperties properties = CraterProperties.getInstance();
             CloseableHttpClient httpclient = HttpClients.createDefault();
```
not closing httpclient connection. Why not use "try clause with resources" for any closeable interface?

angshu commented on feed-integration-webapp/src/main/resources/atomfeed.properties:
```diff
@@ -11,6 +11,6 @@ feed.failedEventMaxRetry=10
 openmrs.auth.uri=http://localhost:8080/openmrs/ws/rest/v1/session
 openmrs.patient.feed.uri=http://localhost:8080/openmrs/ws/atomfeed/patient/recent
 openmrs.user=superman
```
should be externalized

angshu commented on feed-integration-webapp/src/main/resources/atomfeed.properties:
```diff
@@ -11,6 +11,6 @@ feed.failedEventMaxRetry=10
 openmrs.auth.uri=http://localhost:8080/openmrs/ws/rest/v1/session
```
are this properties externalizable? application.properties can be, as spring takes the default name for config location .. but these properties can not be hardcoded

TuhinBeehyv replied:
So actually, there is a separate branch, BAH-1938, where crater-atomfeed has been dockerised. In that branch, there are property file templates which replace the property files, thus its externalised. So for the main branch, we have two options: we can keep the keys in each property file which no value and they get replaced by the template property files or we can delete the property files completely and let docker take care of it while creating the container using the templates. Please let me know which one should be done. 
BAH-1938 also has older source code but dockerfile should be upto date.

gsluthra replied:
I will let @N0-man comment on this. 

My thoughts:
1. We have to merge BAH-1938 into main. Can't have code across branches. 
2. Main branch should have properties files with some defaults, but with the ability to override (templatize).
3. No passwords.
4. Definitely should work in docker-compose / minikube. Need to support on-prem and cloud deployments.

TuhinBeehyv replied:
1. I am working on that now.
2. Templates will override the whole file as is. I will keep some defaults and the ability to override it using templates.
3. Passwords will be removed
4. Project has not been configured for minikube. Will support on-perm and cloud deployments.

TuhinBeehyv replied:
Created a PR for merging docker changes 

https://github.com/Bahmni/crater-atomfeed/pull/4

gsluthra commented:
I suggest we merge the PR (after squashing commits to clear previous history), and open separate tasks for pending refactoring. The overall functionality is working and comments don't seem to be severe/security related. 

The reason I would like to merge this is because 
a. We need this functionality to be tested and demoed. Its been under development & review since June! 
b. Want to track other changes in different PRs/JIRA issues.
c. The code appears to be reasonable enough to be merged and then discussions with INFRA team can happen in terms of deployment configuration, etc. 
d. Would like performance testing to also happen for Bahmni Lite, with crater sync on. 

cc: What do you feel @N0-man and @angshu. Do you agree?

N0-man commented:
> I suggest we merge the PR (after squashing commits to clear previous history), and open separate tasks for pending refactoring. The overall functionality is working and comments don't seem to be severe/security related. 
> 
> cc: What do you feel @N0-man and @angshu. Do you agree?

I agree... this PR is huge and undergoing refactoring since 2 months now - possible smell of piling many changes under same PR... let's merge it and we pick any additional refactoring as part of separate focus PR's

gsluthra commented:
Discussed with @angshu - and we all agreed to squash-and-merge and track comments in new JIRA tickets. Thanks!

vchrombie commented:
Good to see progress here, way to go 🚀
